### PR TITLE
Fix missing sql_header for incremental models

### LINKED ISSUE #2136: set_sql_header works well with 'table', but not 'incremental'
### Describe the bug
This test fails when you change 'table' for 'incremental':

test/integration/022_bigquery_test/models/sql_header_model.sql

### Steps To Reproduce

1. Change 'table' to 'incremental' in sql_header_model.sql
2. `dbt run` works well the first time (creates the table)
3. `dbt run` fails the next time (when doing incremental)

### Expected behavior

The incremental run should bring the same headers - the model will again need the temporal function.

### Screenshots and log output

"Function not found: yes_no_to_boolean at [14:8]"

### System information
**Which database are you using dbt with?**
- [ ] postgres
- [ ] redshift
- [x ] bigquery
- [ ] snowflake
- [ ] other (specify: ____________)

**The output of `dbt --version`:**
0.15.2

### PR DESCRIPTION
**Pull request by drewbanin**
resolves #2136

### Description
This PR adds support for injecting a `sql_header` above `merge` statements generated for incremental models.

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

File: core/dbt/include/global_project/macros/materializations/common/merge.sql [hunk 1]
```diff
@@ -87,6 +90,9 @@
 {% macro default__get_insert_overwrite_merge_sql(target, source, dest_columns, predicates) -%}
     {%- set predicates = [] if predicates is none else [] + predicates -%}
     {%- set dest_cols_csv = get_quoted_csv(dest_columns | map(attribute="name")) -%}
+    {%- set sql_header = config.get('sql_header', none) -%}
+
+    {{ sql_header if sql_header is not none }}
```
Comment (jtcohen6) [inline core/dbt/include/global_project/macros/materializations/common/merge.sql]: I think, if we do this, then the `sql_header` will run twice for "dynamic" (scripting) `insert_overwrite` materializations. (The first time will be when creating the temp table.) I assume that's ok?

**Comment by drewbanin** [comment]
Let's get this merged for 0.16.1 - we should consider whether or not to keep this logic in adapter macros, or move it into materializations (the better long-term answer)

File: core/dbt/include/global_project/macros/materializations/common/merge.sql (hunk 1 above)
Comment (drewbanin) [reply core/dbt/include/global_project/macros/materializations/common/merge.sql:95]: This is a good call. Temporary UDFs can't be defined twice in the same BQ script. Think we're going to need to find an alternative approach here.

I kicked around the idea of making this a materialization-level thing, and while I think that might be a good approach long-term, I don't want to touch every dbt materialization in support of this change. Going to try to figure out a nice way to make this sensible for a very specific use case (insert_overwrite on BigQuery incremental models) and localize the jank as much as possible

**Comment by drewbanin** [comment]
@beckjake FYI merging this before the tests pass - it was a changelog merge conflict

**Pull request merged**